iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp

```diff
@@ -0,0 +1,307 @@
+         "bitwidth is not supported for this type");
+  if (type.isIntOrFloat()) return type.getIntOrFloatBitWidth();
+  auto vecType = type.cast<VectorType>();
+  auto elementType = vecType.getElementType();
+  return elementType.getIntOrFloatBitWidth() * vecType.getNumElements();
+}
+
+namespace {
+/// Analysis memref usages to decide if it should be vectorized. Right now the
+/// logic is to vectorize memref only if it is used by
+/// vectortransfer_read/vectortransfer_write operations.
+class MemRefUsageAnalysis {
+ public:
+  explicit MemRefUsageAnalysis(mlir::Operation *);
+
+  // Return true if the operation should be lowered using operations on
+  // cooperative matrix type.
+  bool vectorizeMemRef(Value v) const { return vectorize.count(v); }
+  bool transferConvert(Operation *v) const { return transferOps.count(v); }
+
+ private:
+  void analyzeFunc(FuncOp funcOp);
+  void analyzeAlloc(AllocOp allocOp);
+  bool allVectorUses(Value v, SmallVector<Operation *, 4> &uses);
+  llvm::DenseSet<Value> vectorize;
+  llvm::DenseSet<Operation *> transferOps;
+};
+
+MemRefUsageAnalysis::MemRefUsageAnalysis(mlir::Operation *op) {
+  op->walk([&](Operation *op) {
+    if (auto func = dyn_cast<FuncOp>(op)) analyzeFunc(func);
+    if (auto alloc = dyn_cast<AllocOp>(op)) analyzeAlloc(alloc);
+  });
+}
+
+void MemRefUsageAnalysis::analyzeFunc(FuncOp funcOp) {
+  for (Value arg : funcOp.getArguments()) {
+    SmallVector<Operation *, 4> vectorUses;
+    if (memRefCanBeVectorized(arg.getType()) &&
+        allVectorUses(arg, vectorUses)) {
+      vectorize.insert(arg);
+      transferOps.insert(vectorUses.begin(), vectorUses.end());
+    }
+  }
+}
+
+bool MemRefUsageAnalysis::allVectorUses(Value v,
+                                        SmallVector<Operation *, 4> &uses) {
+  for (Operation *userOp : v.getUsers()) {
+    // Ignore dealloc ops as they translate to noop.
+    if (isa<DeallocOp>(userOp)) continue;
+    // Only vectorize memref used by vector transfer ops.
+    if (!isa<vector::TransferReadOp, vector::TransferWriteOp>(userOp))
+      return false;
+    uses.push_back(userOp);
+  }
+  return true;
+}
+
+void MemRefUsageAnalysis::analyzeAlloc(AllocOp allocOp) {
+  SmallVector<Operation *, 4> vectorUses;
+  if (memRefCanBeVectorized(allocOp.getType()) &&
+      allVectorUses(allocOp, vectorUses)) {
+    vectorize.insert(allocOp);
+    transferOps.insert(vectorUses.begin(), vectorUses.end());
+  }
+}
+
+template <typename OpTy>
+class MemRefConversionPattern : public OpConversionPattern<OpTy> {
+ public:
+  MemRefConversionPattern<OpTy>(MLIRContext *context,
+                                const MemRefUsageAnalysis &memrefUsageAnalysis)
+      : OpConversionPattern<OpTy>::OpConversionPattern(context),
+        memrefUsageAnalysis(memrefUsageAnalysis) {}
+
+ protected:
+  const MemRefUsageAnalysis &memrefUsageAnalysis;
+};
+
+class ProcessFuncArg final : public MemRefConversionPattern<FuncOp> {
```

Comment:
This is fine to be here, but within IREE we dont need to handle function arguments right now.
Related, within IREE entry point functions/dispatch functions, the "arguments" are instantiated in the body using 

```
%0 = iree.placeholder ....{binding = ....} .. : memref<....>
```

So this needs to be handled for IREE. Could we add this (and a test for this).